Compute a fixed point for do-while

hermanventer opened the pull request:
Release note: Handle do-while loops with abstract bounds, for loop bodies only involving locals.

Issue: #1174 

This adds a simple fixed point computation for do-while loops only. Currently only loops that updated only local variables and that have no abrupt exits are supported.

This is deliberately as minimal as possible to make review easier. Given the complexity of this feature, thoughtful review would be great.

NTillmann commented on src/methods/widen.js:
```diff
@@ -121,6 +122,22 @@ export class WidenImplementation {
+      if (result instanceof AbstractValue && result.kind === "widening") {
+        let phiNode = b.phiNode;
+        if (phiNode === undefined) {
+          // Create a temporal location for binding and initialize it with its value at the loop start
```
But then you never store it back to `b.phiNode`?

hermanventer replied:
Oops.

NTillmann commented on src/utils/generator.js:
```diff
@@ -197,6 +197,21 @@ export class Generator {
+      buildNode: function([], context: SerializationContext) {
+        let statements = context.serializeGenerator(body);
+        let testId = test.intrinsicName;
+        invariant(testId !== undefined);
```
Not all abstract values have an intrinsic name. Why is this guaranteed?

NTillmann replied:
Related, I suspect you were trying hard to avoid using the regular mechanism to turn values into nodes, as this `test` thingy is something very special and not a regular value. Scary implicit assumptions... I feel like we need to do some more fundamental distinction to avoid chaos down the road.

hermanventer replied:
The intrinsic name requirement is a precondition of emitDoWhileStatement. I would hoist it, but then Flow complains. Repeating it seemed a bit over the top to me. I'll add a comment.

NTillmann commented on src/utils/generator.js:
```diff
@@ -197,6 +197,21 @@ export class Generator {
+        let statements = context.serializeGenerator(body);
+        let testId = test.intrinsicName;
+        invariant(testId !== undefined);
+        statements.push(t.ifStatement(t.identifier(testId), t.continueStatement(), t.breakStatement()));
```
What's wrong with emitting the condition as part of the `while` clause? A comment would be nice if there's a reason.

hermanventer replied:
I'll look at this again.

hermanventer replied:
I'll move it into the while clause.

NTillmann commented on test/serializer/abstract/DoWhile.js:
```diff
@@ -0,0 +1,8 @@
+let i = 0;
+do {
+  i++;
+} while (i < n);
```
I feel like we need more test cases for all the new code...

NTillmann commented on src/evaluators/DoWhileStatement.js:
```diff
@@ -33,34 +34,67 @@ export default function(
+    let exprRef = env.evaluate(test, strictCode);
+    let exprValue = Environment.GetValue(realm, exprRef);
+    invariant(exprValue instanceof AbstractValue);
+    let cond = bodyGenerator.derive(exprValue.types, exprValue.values, [exprValue], ([n]) => n, {
```
Why do you create `cond` wrapping around `exprValue`?

hermanventer replied:
Since cond is the loop exit condition it needs to be temporal.

NTillmann replied:
I don't see how that's needed. There are no side effects involved here. Is there something concrete that goes wrong when you don't `derive`?

NTillmann replied:
I can see how you'd like to get a value with an intrinsic name as that's what `emitDoWhileStatement` assumes, but I don't see a deep reason why we need to give a name to the value.

hermanventer replied:
When I don't force evaluation of the expression inside the loop body, it does not actually get emitted anywhere if I do the seemingly obvious thing of making the test condition an argument of the entry created by emitDoWhileStatement. Perhaps we can came back to this when you are in the office again?

NTillmann replied:
Okay. Please leave behind a comment / TODO in the code to this end.

hermanventer commented:
I've addressed the review feedback and added a bunch of new test cases that provide significantly better code coverage. Not surprisingly a number of bugs came to light. Fixing these resulting in significant code changes.

simonhj commented on src/evaluators/DoWhileStatement.js:
```diff
@@ -33,34 +34,67 @@ export default function(
-    // a. Let stmt be the result of evaluating Statement.
-    let stmt = env.evaluateCompletion(body, strictCode);
-    invariant(stmt instanceof Value || stmt instanceof AbruptCompletion);
+  let resultOrDiagnostic = realm.evaluateWithUndoForDiagnostic(() => {
```
What does diagnostic mean in this context?

hermanventer replied:
If prepack is unable to interpret a loop because of an unknown loop bound, or because of some other problem in the loop body, it will issue a diagnostic and throw a FatalError. This helper function captures the diagnostic, squashes the exception and undoes any effects from the abortive loop interpretation, leaving the field open for the fixed point computation to have a go.

simonhj commented on src/realm.js:
```diff
@@ -453,6 +461,99 @@ export class Realm {
+          return (loopBody(): any);
+        });
+        this.restoreBindings(effects1[2]);
+        this.restoreProperties(effects1[3]);
```
Would be easier to read if parts of the tuple was given names with pattern matching.

hermanventer replied:
I did not think it worthwhile because the functions strongly suggest what they expect their arguments to be and because Flow types make sure that right tuple column is used.

simonhj commented on src/evaluators/DoWhileStatement.js:
```diff
@@ -33,34 +34,67 @@ export default function(
+  let fbody = () => env.evaluateCompletion(body, strictCode);
+  let effects = realm.evaluateForFixpointEffects(ftest, fbody);
+  if (effects !== undefined) {
+    let [outsideEffects, insideEffects] = effects;
```
What is insideEffects here?

hermanventer replied:
The effects that summarizes the state changes of an arbitrary iteration of the loop body. That is used to generate code for the loop body. The outsideEffects is used to summary the entire loop for the subsequent code.

yinghuitan replied:
Maybe worth documenting these two effects in evaluateForFixpointEffects().

simonhj commented on src/realm.js:
```diff
@@ -453,6 +461,99 @@ export class Realm {
+          // effects1 includes every value present in effects2, so doing another iteration using effects2 will not
+          // result in any more values being added to abstract domains and hence a fixpoint has been reached.
+          let [, , bindings1, pbindings1] = effects1;
+          if (pbindings1.size > 0) return undefined;
```
Does `undefined` mean that we cannot reach a fix point?

hermanventer replied:
It means that we give up on trying to reach a fix point. That is a temporary restriction. I plan to deal with properties eventually, but that is a bit harder and will make the pull request too large for review.

NTillmann commented on src/realm.js:
```diff
@@ -453,6 +461,99 @@ export class Realm {
+          invariant(val instanceof AbstractValue);
+          val._buildNode = nodes => (id: any);
+          val.kind = "widened";
+          return t.expressionStatement(t.assignmentExpression("=", id, v));
```
All assignments are emitted in some order. Couldn't the right-hand sides refer to the left-hand sides in some cyclic way, resulting in bad code generation? (I tried playing around for a while, but I couldn't produce a test case that exposes this...)

hermanventer replied:
I'm a bit worried about this too. Let me try and cook up a test case that will clarify things one way or another.

hermanventer replied:
Got a test case that showed up some problems with sequencing. Things have now been refactored so that every rh side is first evaluated into temporary variables before the locals (in the guise of their phi nodes) are updated. Any references to phi nodes inside the loop body will always be to their values at loop iteration start. I.e. everything is done the SSA way until all expressions have been evaluated, then the phi nodes are updated in a batch of assignments that do no expression evaluation.

yinghuitan commented on src/environment.js:
```diff
@@ -93,6 +93,8 @@ export type Binding = {
   name: string,
   isGlobal: boolean,
+  // bindings that are assigned to inside loops with abstract termination conditions need temporal locations
+  phiNode?: AbstractValue,
```
Is phiNode's lifetime spanning across widenBindings() invocation session? If not, since phiNode is only used inside widenBindings(), maybe we could create a Map<Binding, AbstractValue> map inside widenBindings to temporal store these values? By doing this, we do not need to expose "phiNode" outside of widenBindings() function and these values can be released after widenBindings() call.

hermanventer replied:
The point is exactly that the same phi node should be used across all fix point iterations.
